Sources/MongoSwift/ReadWriteConcern.swift

```diff
@@ -114,3 +113,166 @@ extension ReadConcern: Equatable {
+    }
+
+    /// Indicates whether this is the default write concern.
+    public var isDefault: Bool {
+        return mongoc_write_concern_is_default(self._writeConcern)
+    }
+
+    /// Indicates whether the combination of values set on this `WriteConcern` is valid.
+    public var isValid: Bool {
+        return mongoc_write_concern_is_valid(self._writeConcern)
+    }
+
+    /// Initializes a new, empty `WriteConcern`.
+    public init() {
+        self._writeConcern = mongoc_write_concern_new()
+    }
+
+    /// Initializes a new `WriteConcern`.
+    public init(journal: Bool? = nil, w: W? = nil, wtimeoutMS: Int32? = nil) {
+        self._writeConcern = mongoc_write_concern_new()
+        if let journal = journal { mongoc_write_concern_set_journal(self._writeConcern, journal) }
+        if let wtimeoutMS = wtimeoutMS { mongoc_write_concern_set_wtimeout(self._writeConcern, wtimeoutMS) }
+
+        if let w = w {
+            switch w {
+            case let .number(wNumber):
+                mongoc_write_concern_set_w(self._writeConcern, wNumber) 
+            case let .tag(wTag):
+                mongoc_write_concern_set_wtag(self._writeConcern, wTag)
+            case .majority:
+                mongoc_write_concern_set_w(self._writeConcern, MONGOC_WRITE_CONCERN_W_MAJORITY)
+            }
+        }
+    }
+
+    /// Initializes a new `WriteConcern` by copying a `mongoc_write_concern_t`.
+    /// The caller is responsible for freeing the original `mongoc_write_concern_t`.
+    internal init(_ writeConcern: OpaquePointer?) {
+        self._writeConcern = mongoc_write_concern_copy(writeConcern)
+    }
+
+    private enum CodingKeys: String, CodingKey {
+        case w, j, wtimeout
+    }
+
+    public required convenience init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        let w = try container.decodeIfPresent(W.self, forKey: .w)
+        let journal = try container.decodeIfPresent(Bool.self, forKey: .j)
+        let wtimeoutMS = try container.decodeIfPresent(Int32.self, forKey: .wtimeout)
+        self.init(journal: journal, w: w, wtimeoutMS: wtimeoutMS)
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encodeIfPresent(self.w, forKey: .w)
+        try container.encodeIfPresent(self.journal, forKey: .j)
+        try container.encodeIfPresent(self.wtimeoutMS, forKey: .wtimeout)
+    }
+
+    deinit {
+        guard let writeConcern = self._writeConcern else { return }
+        mongoc_write_concern_destroy(writeConcern)
+        self._writeConcern = nil
+    }
+}
+
+/// An extension of `WriteConcern` to make it `CustomStringConvertible`.
+extension WriteConcern: CustomStringConvertible {
+    public var description: String {
+        if let encoded = try? BsonEncoder().encode(self).description {
+            return encoded
+        }
+        return ""
```

Comment:
Should an empty write concern be "{}", or is this return case for a `nil` value?

Reply:
In [Error Handling: Converting Errors to Optional Values](https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html), does `try?` assignment catch and ignore any error thrown -- essentially converting it to a `nil` for assignment? If there is a real error, wouldn't you want to propagate that to the user?

I'm not sure that an error string in place of `""` would be preferable, as the user still has to examine that string and infer that an error occurred.

> if we are certain that the encoding will never fail given a successfully initialized WriteConcern, we could also just trigger a preconditionFailure here

This seems like a reasonable thing to do. If the WriteConcern is valid, we should be able to assert that it can be encoded successfully. Any error during an encode would be seriously unexpected IMO.

Reply:
yes, basically, it ignores the error. which I did under the assumption that there would never actually be a relevant encoding error. however, that would not be true if we check validity at the time of encoding rather than time of initialization. 

if we *were* to throw in the `WriteConcern` initializer for an invalid combination of inputs, we could write this as follows:
```swift
public var description: String {
    do {
        return try BsonEncoder().encode(self).description
    } catch {
        preconditionFailure("Error encoding WriteConcern: \(error)")
    }
}
```

if we throw at encoding time, then we need to do something like
```swift
public var description: String {
    do {
        return try BsonEncoder().encode(self).description
    } catch MongoError.writeConcernError {
        // do whatever want to do here for an invalid WriteConcern
    } catch {
        // this is a very unexpected error since the WriteConcern was valid
        preconditionFailure("Error encoding WriteConcern: \(error)")
    }
}
```

Reply:
I'm not really sure what is better 🤔